sdk/core/core-rest-pipeline/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -32,26 +65,165 @@ export interface BearerTokenAuthenticationPolicyOptions {
+   */
+  function getChallenge(response: PipelineResponse): string | undefined {
+    const challenge = response.headers.get("WWW-Authenticate");
+    if (response.status === 401 && challenge) {
+      return challenge;
+    }
+    return;
+  }
+
   return {
     name: bearerTokenAuthenticationPolicyName,
+    /**
+     * If there's no challenge parameter:
+     * - It will try to retrieve the token using the cache, or the credential's getToken.
+     * - Then it will try the next policy with or without the retrieved token.
+     *
+     * It uses the challenge parameters to:
+     * - Skip a first attempt to get the token from the credential if there's no cached token,
+     *   since it expects the token to be retrievable only after the challenge.
+     * - Prepare the outgoing request if the `prepareRequest` method has been provided.
+     * - Send an initial request to receive the challenge if it fails.
+     * - Process a challenge if the response contains it.
+     * - Retrieve a token with the challenge information, then re-send the request.
+     */
     async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
-      const token = await getToken({
-        abortSignal: request.abortSignal,
-        tracingOptions: request.tracingOptions
-      });
-      request.headers.set("Authorization", `Bearer ${token}`);
-      return next(request);
+      let accessToken = tokenCache.getCachedToken();
+      if (!accessToken) {
+        // retrieves AAD access token
+        const token = await retrieveToken(request);
+        request = assignToken(request, token);
+      } else {
+        request = assignToken(request, accessToken.token);
+      }
+
+      if (callbacks?.prepareRequest) {
+        await callbacks.prepareRequest(request);
+      }
+
+      let challenge: string | undefined;
+      let response: PipelineResponse;
+      try {
+        response = await next(request);
+        challenge = getChallenge(response);
+      } catch (err) {
+        if (err instanceof RestError && err.statusCode === 401) {
+          challenge = getChallenge(err.response!);
+        } else {
+          throw err;
+        }
+      }
+
+      if (challenge && callbacks?.processChallenge) {
+        // processes challenge
+        await callbacks.processChallenge(challenge, request);
+        return next(request);
+      }
+
+      //@ts-ignore assigned in try block
+      return response;
     }
   };
 }
+
+type ValidParsedWWWAuthenticateProperties =
+  // "authorization_uri" was used in the track 1 version of KeyVault.
+  // This is not a relevant property anymore, since the service is consistently answering with "authorization".
+  // | "authorization_uri"
+  | "authorization"
+  | "claims"
+  // Even though the service is moving to "scope", both "resource" and "scope" should be supported.
+  | "resource"
+  | "scope"
+  | "service";
```

Comment:
@sadasant I added some more entries.  I wonder whether there's a spec that we can find all supported names.

Reply:
These are fine for now. We can add more on demand later.